### ISSUE TITLE
Enhancement: configuration option to error on unsupported TypeScript version (`onUnsupportedTypeScriptVersion=error`)

### ISSUE THREAD

**Issue by mluypaert** [typescript-eslint/typescript-eslint#10092]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-eslint

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

When using an unsupported TypeScript version, my automated testing should test for this, and fail, as using an unsupported TypeScript version can cause unexpected and unwanted behavior.

The current `warnOnUnsupportedTypeScriptVersion` behavior (when set to `true`) is to display a warning message, but not fail the testing. While useful for testing new typescript version, this is not desired behaviour in a production environment, where I want my (style) tests to be reliable and updates to typescript to be testable in an safe and automated way. The output of these ESlint tests are not inspected manually, and only errors will be detected.

My solution proposal would be to update the `warnOnUnsupportedTypeScriptVersion` configuration option to become `onUnsupportedTypeScriptVersion` instead, with possible values
 * `warn` to only display unsupported typescript version message (the current default, equal to `warnOnUnsupportedTypeScriptVersion=true`)
 * `error` to display unsupported typescript version message and fail/error (new behavior)
 * `ignore` to not display anything on usage of unsupported typescript version and just run the tests as if the typescript version would be supported (equal to `warnOnUnsupportedTypeScriptVersion=false`).

I would further argue that given the above solution proposal, `error` should be the default behaviour, as it is safer than `warn` and `ignore`.

**Comment by bradzacher** [comment]
Generally speaking -- each new version of TS will either "just work" or it will crash.
It's a very, very rare case that it will produce "undefined behaviour".

> I want my (style) tests to be reliable and updates to typescript to be testable in an safe and automated way

Could you expand on why you want the parser to specifically crash?

**Comment by mluypaert** [comment]
> Generally speaking -- each new version of TS will either "just work" or it will crash. It's a very, very rare case that it will produce "undefined behaviour".
> 
> > I want my (style) tests to be reliable and updates to typescript to be testable in an safe and automated way
> 
> Could you expand on why you want the parser to specifically crash?

Sure, the answer to why I want the parser to throw an exit code indicating error rather than merely printing a warning and exiting with an exit code that indicates success is easy: CI/CD.

As part of my CI/CD, I run my eslint (through [next.js](https://nextjs.org/docs/app/building-your-application/configuring/eslint)) on every PR submitted to my repository, and require it to pass before allowing merging of said PR.
 * When the ESLint parser returns an exit code indicating error, this check fails and the submitter will investigate the logs and notice the used typescript version is not compatible (which means the submitter will need to update the frozen TS version to remain frozen on a lower, compatible version).
 * When the ESLint parser returns a success exit code (0) and only prints a warning message, no-one ever notices this incompatibility, and the incompatible TS version upgrade will get merged in, possibly breaking tests elsewhere.

This PR-validation is also applied to version upgrade proposals from my dependency manager (dependabot), which means when the PR-validation checks pass (including eslint) the assumption is this version upgrade can safely be applied. This is currently not the case for ESLint and Typescript, because of this warning behaviour on incompatible versions.

**Comment by bradzacher** [comment]
I guess my point is what I stated above:

> Generally speaking -- each new version of TS will either "just work" or it will crash.

The times that things fall into an "undefined behaviour" state from a TS update are truly 1-in-a-million -- so it's safe to just disable the log -- a lot of people do.

---

Personally I don't know if adding a hard failure option is necessarily right -- this behaviour serves as a way for us to loudly warn about our "peer dependency" requirement.

The reason we have that warning log is primarily so that people don't report issues saying things don't work on the new TS version. People used to do that and it was really frustrating for us -- they'd upgrade to the beta and RC versions and then file issues reporting things.

But realistically it's the same level of enforcement as declaring a peer dependency range -- which would just log a warning to the console during the install. Which is why I'm hesitant to go further to crash things.

**Comment by kirkwaiblinger** [comment]
I'm +0.5. I get really annoyed when I can't make a CLI tool error when it has warnings/errors. I get that it's probably a nonissue in most cases here but I'd still rather explicitly opt into the unsupported version if I need to than do so unknowingly.

**Comment by JoshuaKGoldberg** [comment]
Agreed, I'm +1 on this. I'm constantly getting annoyed at Dependabot/Renovate/etc. merging things with warnings. This very issue is one I've faced before and been annoyed at.

**Comment by bradzacher** [comment]
> I'm constantly getting annoyed at Dependabot/Renovate/etc. merging things with warnings

What I'm saying though is that the bots wouldn't respect a peer dep range.
If we declared `peerDependency: { typescript: '>=4.8.4 <5.6.0' }` then dependabot would not respect that and it would just bump the TS version and CI would pass.

In essence that's what we're declaring with our log - we're just doing it louder than the package manager would.

That's why I'm -1 on this -- cos we'd be doing something pretty unique amongst the JS ecosystem.

**Comment by mluypaert** [comment]
> > I'm constantly getting annoyed at Dependabot/Renovate/etc. merging things with warnings
> 
> What I'm saying though is that the bots wouldn't respect a peer dep range. If we declared `peerDependency: { typescript: '>=4.8.4 <5.6.0' }` then dependabot would not respect that and it would just bump the TS version and CI would pass.
> 
> In essence that's what we're declaring with our log - we're just doing it louder than the package manager would.
> 
> That's why I'm -1 on this -- cos we'd be doing something pretty unique amongst the JS ecosystem.

My proposal is not to declare this in the peer dependencies (although you could probably do that in addition), but rather to have an option to fail ESLint runs when an unsupported typescript version is detected. That way even if dependabot would ignore the peerdependencies and install a typescript update that is not compatible, running eslint would reveal this incompatibility by failing the test, rather only printing to the logs.
As mentioned earlier, logs are only useful to humans, not to automated systems (CI/CD pipelines) that rely on exit codes for error detection.

**Comment by bradzacher** [comment]
Sorry - to clarify - my example of peer dependencies was to highlight that our warning log right now is just "peer dependencies but loud". The reason we have a warning log instead of a peer dependency is so that we can be louder to people (for the reasons I mentioned above) - but ultimately `typescript` is still really a peer dependency for us.

The point I was making was that if we never made the change to add this log and just used peer deps - then your dependabot would still merge a TS version bump as CI would pass. 

The point I'm getting at is that crashing at runtime due to an incorrect peer dependency range that would likely otherwise work just fine would be something very unique amongst the NodeJS ecosystem.

**Comment by ronami** [comment]
Since there's already a warning for this, I'm +0.5 on adding the option to exit with status code 1 instead. 

Warnings are often ignored, and being able to crash the process immediately instead makes sense to me (even if this doesn't practically happen usually).

**Comment by JoshuaKGoldberg** [comment]
I'm also in favor of this, for that reason. It's unfortunate that package managers and Dependabot/Renovate/etc. don't give better controls out-of-the-box to help us out here. But I've been bit by this myself several times since the issue was filed.

Marking as accepted since we're net +1.5 in favor.
